device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java

```diff
@@ -0,0 +1,211 @@
+        String firstLine = connectResponseLines[connectResponseStart];
+        if (!firstLine.startsWith(HTTP_VERSION_1_1))
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d.  Expected first response line to start with %s, but proxy returns \"%s\"", host, port, HTTP_VERSION_1_1, firstLine));
+        }
+
+        String[] replyStrParts = firstLine.split(" ");
+        if (replyStrParts.length < 2)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to contain a space between http version and status code, but was %s", host, port, firstLine));
+        }
+
+        int connectResponseStatusCode;
+        try
+        {
+            connectResponseStatusCode = Integer.parseInt(replyStrParts[1]);
+        }
+        catch (NumberFormatException e)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to contain a status code but status code could not be parsed. Response was %s", host, port, firstLine));
+        }
+
+        if (connectResponseStatusCode <= 199 || connectResponseStatusCode >= 300)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to return status code 2XX but status code was %d", host, port, connectResponseStatusCode));
+        }
+
+        log.trace("HTTP proxy responded to connect request with status {}, so the proxy connect was successful", connectResponseStatusCode);
+    }
+
+    private interface ProxiedSSLSocketNonDelegatedFunctions
+    {
+        void connect(SocketAddress socketAddress, int timeout) throws IOException;
+        void connect(SocketAddress socketAddress) throws IOException;
+        void close() throws IOException;
+    }
+
+    @RequiredArgsConstructor
+    class HttpConnectResponseReader
+    {
+        private boolean alreadyRead = false;
+        @NonNull private InputStream inputStream;
+        @NonNull private Charset byteEncoding;
+
+        String readHttpConnectResponse() throws IOException
+        {
+            if (alreadyRead)
+            {
+                throw new IOException("Http connect response has already been read");
+            }
+
+            ByteArrayOutputStream httpLineOutputStream = new ByteArrayOutputStream();
+            int consecutiveHttpNewlines = 0;
+            int previousCharacter = -1;
+            while (consecutiveHttpNewlines < 2)
+            {
+                int i = inputStream.read();
+                if (i == -1)
+                {
+                    inputStream.close();
+                    throw new IOException("Unexpected EOF from proxy");
+                }
+
+                httpLineOutputStream.write((byte) i);
+
+                if (i == '\n')
```

Comment:
What if `\r\n\r\r\r\n`?

Reply:
Okay, I've added logic to catch that, and also to catch ```\r\n\n\r\n```

Reply:
How about:
```
if (i == '\n' && previousCharacter == '\r') {
  consecutiveHttpNewlines++;
} else if (i != '\r'||previousCharacter == '\n' ) {
  consecutiveHttpNewlines=0;
}
```

Reply:
Now, I'll just keep a char[] of the last 4 read chars, and loop until they match ```\r\n\r\n```